std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,576 @@
+    //         }
+    //     }
+    // }
+
+    array::new(100, |i| active_row * (preimage[i] - preimage[i]') = 0);
+
+    // // C'[x, z] = xor(C[x, z], C[x - 1, z], C[x + 1, z - 1]).
+    // for x in 0..5 {
+    //     for z in 0..64 {
+    //         builder.assert_bool(local.c[x][z]);
+    //         let xor = xor3_gen::<AB::Expr>(
+    //             local.c[x][z].into(),
+    //             local.c[(x + 4) % 5][z].into(),
+    //             local.c[(x + 1) % 5][(z + 63) % 64].into(),
+    //         );
+    //         let c_prime = local.c_prime[x][z];
+    //         builder.assert_eq(c_prime, xor);
+    //     }
+    // }
+
+    array::map(c, |i| force_bool(i));
+    let xor: expr, expr -> expr = |a, b| a + b - 2*a*b;
+    let xor3: expr, expr, expr -> expr = |a, b, c| a + b + c - 2*(a*b + b*c + c*a) + 4*a*b*c;
+    // a b c xor3
+    // 0 0 0  0
+    // 0 0 1  1
+    // 0 1 0  1
+    // 0 1 1  0
+    // 1 0 0  1
+    // 1 0 1  0
+    // 1 1 0  0
+    // 1 1 1  1
+    array::new(320, |i| {
+        let x = i / 64;
+        let z = i % 64;
+        xor3(c[i], c[((x + 4) % 5) * 64 + z], c[((x + 1) % 5) * 64 + ((z + 63) % 64)]) - c_prime[i] = 0
+    });
+
+    // // Check that the input limbs are consistent with A' and D.
+    // // A[x, y, z] = xor(A'[x, y, z], D[x, y, z])
+    // //            = xor(A'[x, y, z], C[x - 1, z], C[x + 1, z - 1])
+    // //            = xor(A'[x, y, z], C[x, z], C'[x, z]).
+    // // The last step is valid based on the identity we checked above.
+    // // It isn't required, but makes this check a bit cleaner.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         let get_bit = |z| {
+    //             let a_prime: AB::Var = local.a_prime[y][x][z];
+    //             let c: AB::Var = local.c[x][z];
+    //             let c_prime: AB::Var = local.c_prime[x][z];
+    //             xor3_gen::<AB::Expr>(a_prime.into(), c.into(), c_prime.into())
+    //         };
+
+    //         for limb in 0..U64_LIMBS {
+    //             let a_limb = local.a[y][x][limb];
+    //             let computed_limb = (limb * BITS_PER_LIMB..(limb + 1) * BITS_PER_LIMB) // bigger address correspond to more significant bit
+    //                 .rev()
+    //                 .fold(AB::Expr::zero(), |acc, z| {
+    //                     builder.assert_bool(local.a_prime[y][x][z]);
+    //                     acc.double() + get_bit(z)
+    //                 });
+    //             builder.assert_eq(computed_limb, a_limb);
+    //         }
+    //     }
+    // }
+
+    array::map(a_prime, |i| force_bool(i));
+    array::new(100, |i| {
+        let y = i / 20;
+        let x = (i / 4) % 5;
+        let limb = i % 4;
+        let get_bit: int -> expr = |z| xor3(a_prime[y * 320 + x * 64 + z], c[x * 64 + z], c_prime[x * 64 + z]);
+        let computed_limb: expr = utils::fold(16, |z| get_bit((limb + 1) * 16 - 1 - z), 0, |acc, e| (acc * 2 + e));
```

Comment:
What about
```suggestion
        array::reverse(array::new(16, |z| get_bit((limb + 1) * 16 + z)))
```
?
(and add `reverse` to `std::array`)

Reply:
Done and integrated